### ISSUE TITLE
LLVM 8.0.0 beta string parameters are *sbyte

### ISSUE THREAD

**Issue by yowl** [microsoft/LLVMSharp#123]
Hi,  I know its beta, but is the plan to allow strings to be passed to LLVM functions without converting to *sbyte.  In 5.0.0 this was possible:

```
                LLVMBasicBlockRef trapBlock = LLVM.AppendBasicBlock(_llvmFunction, "Trap");
```
In 8.0.0, the parameter is a `sbyte*`.  Plus most of the functions now return pointers meaning that using classes need to be `unsafe`.  Is that by desgin?

**Comment by yowl** [comment]
I see that the OOP approach is ok, i.e. that line can be instead:
```
                LLVMBasicBlockRef trapBlock = _llvmFunction.AppendBasicBlock("Trap");
```
Is the OOP approach the way forward?

**Comment by tannergooding** [comment]
> Is the OOP approach the way forward?

The short answer: Yes.

Automatically generated interop bindings really need to be a 1-to-1 mapping as there are too many variables that come into play otherwise. It is almost impossible, especially when cross-platform comes into play, to correctly determine if an arbitrary `char *` on the native side should be a `sbyte[]`, a `string`, a `ref sbyte`, an `out sbyte`, or some other binding on the managed side. Having a non 1-to-1 mapping also adds some level of overhead that may not be desirable in all scenarios.

As such, the "raw" bindings are now 1-to-1 mappings with the underlying native definitions. These have been moved into the `LLVMSharp.Interop` namespace for the 9.0.0 beta.

There then exists a minimal abstraction that makes these APIs "safe" to use from managed code. These are exposed as "instance" methods off the strongly typed handles. An example is `LLVMContextRef` which is implicitly convertible from the `LLVMOpaqueContext*` used by the raw bindings.

There are plans to then expose nicer APIs that more closely mirror the LLVM C++ API surface. We've already exposed such bindings for ClangSharp if you would like an example of how that will be done.
